coderd/workspaceagents.go

```diff
@@ -1835,22 +1835,28 @@ func (api *API) workspaceAgentUpdateMetadata(ctx context.Context, workspaceAgent
 // @Router /workspaceagents/{workspaceagent}/watch-metadata [get]
 // @x-apidocgen {"skip": true}
 func (api *API) watchWorkspaceAgentMetadata(rw http.ResponseWriter, r *http.Request) {
-	var (
-		ctx            = r.Context()
-		workspaceAgent = httpmw.WorkspaceAgentParam(r)
-		log            = api.Logger.Named("workspace_metadata_watcher").With(
-			slog.F("workspace_agent_id", workspaceAgent.ID),
-		)
+	// Allow us to interrupt watch via cancel.
+	ctx, cancel := context.WithCancel(r.Context())
+	defer cancel()
+	r = r.WithContext(ctx) // Rewire context for SSE cancellation.
+
+	workspaceAgent := httpmw.WorkspaceAgentParam(r)
+	log := api.Logger.Named("workspace_metadata_watcher").With(
+		slog.F("workspace_agent_id", workspaceAgent.ID),
 	)
 
 	// Send metadata on updates, we must ensure subscription before sending
 	// initial metadata to guarantee that events in-between are not missed.
 	update := make(chan workspaceAgentMetadataChannelPayload, 1)
 	cancelSub, err := api.Pubsub.Subscribe(watchWorkspaceAgentMetadataChannel(workspaceAgent.ID), func(_ context.Context, byt []byte) {
```

Comment:
Is the `func(context.Context, []byte)` we pass to `api.Pubsub.Suscribe` getting ripe for extraction?

Reply:
Could you clarify? I'm not sure I understand the question.

Reply:
I mean, should we refactor the second argument of this function into its own function as opposed to leaving it anonymous? Not in this PR of course.

Reply:
Ah, I don't think that's necessary, it's very short and the only logic in it is the slice merger. We could break that out into its own function, though.